src/platform/w3dengine/client/w3ddisplay.cpp

```diff
@@ -60,80 +94,457 @@ W3DDisplay::W3DDisplay() :
+                    // Calls some debug stuff in WB
+                    g_theMouse->Draw();
+                }
+
+                if (m_videoStream) {
+                    if (m_videoBuffer) {
+                        Draw_VideoBuffer(m_videoBuffer, 0, 0, Get_Width(), Get_Height());
+                    }
+                }
+
+                if (m_unkDisplayString) {
+                    int x;
+                    int y;
+                    m_unkDisplayString->Get_Size(&x, &y);
+                    m_unkDisplayString->Draw((Get_Width() / 2) - x / 2, Get_Height() - y - 20, 0xFF000000, 0, 0, 0);
+                }
+
+                Render_LetterBox(now);
+
+                if (m_cinematicText != Utf8String::s_emptyString) {
+                    if (m_cinematicTextFrames) {
+                        DisplayString *d = g_theDisplayStringManager->New_Display_String();
+                        d->Set_Word_Wrap(g_theDisplay->Get_Width() - 20);
+                        d->Set_Word_Wrap_Centered(true);
+                        Utf16String str;
+                        str.Translate(m_cinematicText);
+                        d->Set_Text(str);
+                        d->Set_Font(m_cinematicFont);
+                        int h = (int)((float)g_theDisplay->Get_Height() * 0.9f);
+                        int w;
+
+                        if (d->Get_Width(-1) <= g_theDisplay->Get_Width()) {
+                            w = (g_theDisplay->Get_Width() - d->Get_Width(-1)) / 2;
+                        } else {
+                            w = 20;
+                        }
+
+                        d->Draw(w, h, 0xFFFFFFFF, 0);
+                        m_cinematicTextFrames--;
+                    }
+                }
+
+                if (m_debugDisplayCallback) {
+                    Draw_Current_Debug_Display();
+                }
+
+#ifdef GAME_DEBUG_STRUCTS
+                if (g_theWriteableGlobalData->m_benchmarkTimer) {
+                    Draw_Benchmark();
+                }
+#endif
+                if (g_theWriteableGlobalData->m_showFrameRateBar) {
+                    Draw_Graphical_Framerate_Bar();
+                }
+
+                W3D::End_Render(true);
+            l2:
+                if (g_theScriptEngine->Is_Time_Frozen_Debug() || g_theScriptEngine->Is_Time_Frozen_Script()
+                    || g_theGameLogic->Is_Game_Paused()) {
+                    b = false;
+                }
+                goto l3;
+            }
+
+            g_theInGameUI->Draw();
+
+            if (g_theMouse) {
+                g_theMouse->Draw();
+            }
+
+            W3D::End_Render(true);
+        l3:
+            if (!b || g_theTacticalView->Is_Camera_Movement_Finished()) {
+                return;
+            }
+        }
+
+        if (couldRender) {
+            couldRender = false;
+            captainslog_debug("Could not do WW3D::Begin_Render()!  Are we ALT-Tabbed out?\n");
+        }
+
+        goto l2;
```

Comment:
Are these goto's as per original code or does just compiler or decompiler generate these? This function is very difficult to read.

Reply:
decompiler but it's not always that simple to unscramble these, in such situations a TODO to do it later is acceptable, this by far won't be the last of them

Reply:
fixed.

Reply:
Labels or gotos still need TODO cleanup comments